Model Registry: Register New Version form

mturley opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- https://issues.redhat.com/browse/RHOAIENG-123456 -->

Closes [RHOAIENG-7571](https://issues.redhat.com/browse/RHOAIENG-7571)

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

Implements the "Register New Version" button in these places:
* RegisteredModelsTableToolbar (no specific registered model)
* ModelVersionListView empty state (under a specific registered model)
* ModelVersionListView toolbar (under a specific registered model)

Depending on where the action was selected, the user will be redirected to one of these routes which both render the RegisterVersion page:
* `/modelRegistry/{mrName}/registerVersion` - Presents a typeahead select for choosing a model
* `/modelRegistry/{mrName}/registeredModels/{registeredModelId}/registerVersion` - Pre-selects the model and disables the model field

The RegisterVersion form is mostly a subset of the RegisterModel form - Version Details and Model Location fields in both are factored out into a new component `RegistrationCommonFormSections`. The only difference is that instead of entering a model name and description we let the user select a model (or preselect it for them), and then we show some help text and prefill some fields based on the selected model.

When a model is selected, the hook `usePrefillRegisterVersionFields` re-renders with a new `registeredModel` value which causes it to fetch the versions of that model, then fetch the artifacts for that version, finally deriving a `latestVersion` and `latestArtifact`. The `useEffect` in that hook is then triggered to update form values based on `latestArtifact` (source model format/version and model location endpoint/bucket/region), and `latestVersion` is returned and passed into `RegistrationCommonFormSections` where it is used to show the "Current version is ..." help text.

If the model selection changes and the selected model has no version/artifact, nothing will be prefilled. If it has a version and artifact but the URI on that artifact is malformed (cannot be parsed into endpoint/bucket/region), only the source model format/version is prefilled. Any fields that can't be prefilled are reset to empty, so switching between models should not leave behind values prefilled from another model.

Registering a new version from the registered models table (no preselected model):

![Screenshot 2024-08-15 at 11 25 31 AM](https://github.com/user-attachments/assets/f53351d1-4104-481c-bfa8-377d065238a4)

![Screenshot 2024-08-15 at 11 25 50 AM](https://github.com/user-attachments/assets/bed21f1c-e13e-420f-a5fa-b3f35f77ba5d)

Selecting a model prefills fields:

![Screenshot 2024-08-15 at 11 26 07 AM](https://github.com/user-attachments/assets/3adcbad9-09d9-4fcf-9eab-2573aad0ee37)

Registering a new version from a registered model's page (when there are no versions):

![Screenshot 2024-08-15 at 11 27 47 AM](https://github.com/user-attachments/assets/8ddd2c6c-6c14-4b26-9faf-7ee13f6805ff)

Registering a new version from a registered model's page (when there are already versions):

![Screenshot 2024-08-15 at 11 26 51 AM](https://github.com/user-attachments/assets/fffc4615-a708-4295-bfee-9642a44cec47)

Model is preselected and fields are already filled:

![Screenshot 2024-08-15 at 11 27 10 AM](https://github.com/user-attachments/assets/90b78bc0-eaf1-4992-b67d-c3d8b1e263a9)

cc @yih-wang 


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

Tested on modelregistry-ui cluster. Tried registering versions by reaching the form from all 3 buttons, covered all cases described above re: prefilling.

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->

Unit tests added for new `getLastCreatedItem` util, and tests moved for utils that moved (`filterArchiveVersions`, `filterLiveVersions`).

Cypress tests added to cover all cases of prefilling/clearing behavior, form validation, and creating expected resources on submit in each case.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

manaswinidas reviewed (changes requested):
When I'm trying to `Register new version` via dropdown menu in the Registered models table.. after selecting model name.. the model location fields are pre-filled but in case of `Register new version` button above the Versions table ... although the model name is prefilled, the model location fields aren't pre-filled, is this expected?
<img width="916" alt="Screenshot 2024-08-12 at 6 14 29 PM" src="https://github.com/user-attachments/assets/36a2a85a-1c43-49f8-b303-93e46b1fa18d">

mturley commented:
nope, that is not expected... thanks, good catch.

yih-wang commented:
Thanks for all the work @mturley! The only comment I have is regarding the visual style of the read-only fields (model registry and model name). I know you are following the mockup to use the read-only status for a text field. But recently I just noticed the other forms are using plain text to show the read-only values. See the image below. I think we should follow that consistency. I'm fine to both address it in this issue or leave it to post-mvp. Sorry for the duplicated work!
<img width="547" alt="image" src="https://github.com/user-attachments/assets/e5fff734-e702-442e-810b-ed5125e38f1d">

manaswinidas reviewed (commented):
The registration and auto-fill work fine now. But after a successful registration the redirect URL should be the Model version table(according to mocks). It's going to the version Details page for me. Let me know if I'm not referring to the latest mocks.
<img width="988" alt="Screenshot 2024-08-16 at 6 17 39 PM" src="https://github.com/user-attachments/assets/e56d5ab5-e187-433a-b123-9c7d862623a5">

mturley commented:
@manaswinidas good catch, I've rebased and fixed that. I had to squash the commits to avoid a bunch of pain in rebasing on top of https://github.com/opendatahub-io/odh-dashboard/pull/3084, sorry if that makes it harder to re-review -- mostly I moved that PR's changes to RegisterModel into RegistrationCommonFormSections.

@yih-wang no worries! yeah, do you mind including that somewhere we're tracking post-MVP refactors? I'm not sure the best place for it for now - we need to meet and follow up to make sure there are ENG issues opened for all those under https://issues.redhat.com/browse/RHOAIENG-10405 at some point soon.

gitdallas commented on frontend/src/concepts/modelRegistry/utils.ts:
```diff
@@ -37,3 +39,27 @@ export const uriToObjectStorageFields = (uri: string): ObjectStorageFields | nul
+    },
+  )[0];
+
+export const filterArchiveVersions = (modelVersions: ModelVersion[]): ModelVersion[] =>
```
could combine these to 2 by passing ModelState or even to 1 if you didn't mind doing `models: ModelVersion[] | RegisteredModel[]`, doesn't matter though

gitdallas replied:
i see these were just moved from another place anyway 😆

mturley replied:
Yeah 😆 that's fair.. probably better to factor them out using generics or use `ModelRegistryBase`, but that's tech debt for another time I think (or fine as-is if we never get to it)

yih-wang commented:
> @yih-wang no worries! yeah, do you mind including that somewhere we're tracking post-MVP refactors? I'm not sure the best place for it for now - we need to meet and follow up to make sure there are ENG issues opened for all those under https://issues.redhat.com/browse/RHOAIENG-10405 at some point soon.

Add a ux issue [RHOAIUX-317](https://issues.redhat.com/browse/RHOAIUX-317) to track the following work
